llvm/tools/repo2obj/repo2obj.cpp

```diff
@@ -395,7 +395,10 @@ int main(int argc, char *argv[]) {
                 OutputSection<ELFT>::SectionInfo{});
 
       auto const IsLinkOnce =
-          CM.linkage == pstore::repo::linkage_type::linkonce;
+          CM.linkage == pstore::repo::linkage_type::link_once_any ||
+          CM.linkage == pstore::repo::linkage_type::link_once_odr ||
+          CM.linkage == pstore::repo::linkage_type::weak_any ||
```

Comment:
We’re now exposing repo2obj to the idea of weak linkage for the first time: a value of weak_any linkage isn't linkonce even though a linkonce symbol is (in ELF) always weak. This variable, and the logic around it will need to be revised to reflect the difference between “weak” and “linkonce” symbols: should weak be excluded from this collection?

Reply:
Before the changes:

1. The four LLVM linkage types (`LinkOnceAnyLinkage`, `LinkOnceODRLinkage`, `WeakAnyLinkage` and `WeakODRLinkage`) have been mapped to the Repo `linkonce` linkage type. 
2. auto const IsLinkOnce = CM.linkage == pstore::repo::linkage_type::linkonce;
    Therefore, the `IsLinkOnce` flag is true for four LLVM linkage types mentioned in 1.

When I replicated all the LLVM linkage types in the Repo, I tried to keep the same behavior as before. I gives four Repo linkage type here. I am not sure that weak could be excluded from this collection.

I will remove the weak linkage and test using the LLVM project build.

Reply:
Yes, the weak linkages have been excluded from this collection.

Reply:
> When I replicated all the LLVM linkage types in the Repo, I tried to keep the same behavior as before.

Yes I can see that: the behavior is unchanged if you look at the ELF files being produced by repo2obj. However, the database content is now different. Since the compiler is no longer collapsing its internal linkage types into a simpler subset, it’s now up to the downstream tools (repo2obj and the linker ATM) to handle those new types.

We’re introducing new “weak” linkages to the repository, so r2o should probably be generating weak ELF symbols in response.

Reply:
Thanks for explaining this. 
After excluding the weak linkage, the LLVM project passed build and test.

Reply:
> After excluding the weak linkage, the LLVM project passed build and test

That’s great, but that simply suggests that the LLVM source doesn’t use weak symbols.

Reply:
> 
>We’re introducing new “weak” linkages to the repository, so r2o should probably be generating weak ELF symbols in response
> 
> That’s great, but that simply suggests that the LLVM source doesn’t use weak symbols.

Sorry, I didn't answer your previous question clearly. the r2o generated the weak symbol in the `linkageToELFBinding` function.

```
template <typename ELFT>
unsigned SymbolTable<ELFT>::linkageToELFBinding(pstore::repo::linkage_type L) {
  switch (L) {
  case pstore::repo::linkage_type::internal_no_symbol:
  case pstore::repo::linkage_type::internal:
    return llvm::ELF::STB_LOCAL;
  case pstore::repo::linkage_type::link_once_any:
  case pstore::repo::linkage_type::link_once_odr:
  case pstore::repo::linkage_type::weak_any:
  case pstore::repo::linkage_type::weak_odr:
    return llvm::ELF::STB_WEAK;
  default:
    return llvm::ELF::STB_GLOBAL;
  }
}
```